PreCheck/doxygen-man/man3/SH_User.3
.TH "SH_User" 3 "Mardi Juillet 2 2013" "Version 0.4" "PreCheck" \" -*- nroff -*-
.ad l
.nh
.SH NAME
SH_User \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <SH_User\&.h>\fP
.PP
Est dérivée de \fBQObject\fP\&.
.PP
Dérivée par \fBSH_Trainee\fP\&.
.SS "Connecteurs publics"

.in +1c
.ti -1c
.RI "static QVariant \fBexists\fP (QVariant login)"
.br
.in -1c
.SS "Signaux"

.in +1c
.ti -1c
.RI "void \fBnameChanged\fP ()"
.br
.ti -1c
.RI "void \fBrolesChanged\fP ()"
.br
.ti -1c
.RI "void \fBvalidityChanged\fP ()"
.br
.in -1c
.SS "Fonctions membres publiques"

.in +1c
.ti -1c
.RI "\fBSH_User\fP (QString \fBname\fP='', int \fBid\fP=0, bool \fBisReceptionist\fP=false, bool \fBisManagerX\fP=false, bool \fBisManagerZ\fP=false, bool \fBisAdministrator\fP=false, \fBQObject\fP *parent=0)"
.br
.ti -1c
.RI "int \fBid\fP () const "
.br
.ti -1c
.RI "bool \fBisAdministrator\fP () const "
.br
.ti -1c
.RI "bool \fBisManagerX\fP () const "
.br
.ti -1c
.RI "bool \fBisManagerZ\fP () const "
.br
.ti -1c
.RI "bool \fBisReceptionist\fP () const "
.br
.ti -1c
.RI "bool \fBisValid\fP () const "
.br
.ti -1c
.RI "QString \fBname\fP () const "
.br
.ti -1c
.RI "int \fBroles\fP () const "
.br
.ti -1c
.RI "virtual bool \fBsave\fP (QString password)"
.br
.ti -1c
.RI "bool \fBsetNewPassword\fP (QString newPass)"
.br
.in -1c
.SS "Fonctions membres publiques statiques"

.in +1c
.ti -1c
.RI "static \fBSH_User\fP * \fBlogIn\fP (QString login, QString pass)"
.br
.ti -1c
.RI "static bool \fBtraineeExists\fP (QString login)"
.br
.ti -1c
.RI "static bool \fBuserExists\fP (QString login)"
.br
.in -1c
.SS "Propriétés"

.in +1c
.ti -1c
.RI "bool \fBadministrator\fP"
.br
.ti -1c
.RI "int \fBid\fP"
.br
.ti -1c
.RI "bool \fBmanagerX\fP"
.br
.ti -1c
.RI "bool \fBmanagerZ\fP"
.br
.ti -1c
.RI "QString \fBname\fP"
.br
.ti -1c
.RI "bool \fBreceptionist\fP"
.br
.ti -1c
.RI "int \fBroles\fP"
.br
.ti -1c
.RI "bool \fBvalid\fP"
.br
.in -1c
.SS "Fonctions membres privées"

.in +1c
.ti -1c
.RI "void \fBsetID\fP (int \fBid\fP)"
.br
.ti -1c
.RI "void \fBsetName\fP (QString \fBname\fP)"
.br
.in -1c
.SS "Attributs privés"

.in +1c
.ti -1c
.RI "bool \fBm_administrator\fP"
.br
.RI "\fIm_administrator \fP"
.ti -1c
.RI "int \fBm_id\fP"
.br
.RI "\fIm_id \fP"
.ti -1c
.RI "bool \fBm_managerX\fP"
.br
.RI "\fIm_managerX \fP"
.ti -1c
.RI "bool \fBm_managerZ\fP"
.br
.RI "\fIm_managerZ \fP"
.ti -1c
.RI "QString \fBm_name\fP"
.br
.RI "\fIm_name \fP"
.ti -1c
.RI "bool \fBm_receptionist\fP"
.br
.RI "\fIm_receptionist \fP"
.in -1c
.SH "Description détaillée"
.PP 
Définition à la ligne 15 du fichier SH_User\&.h\&.
.SH "Documentation des constructeurs et destructeur"
.PP 
.SS "SH_User::SH_User (QStringname = \fC''\fP, intid = \fC0\fP, boolisReceptionist = \fCfalse\fP, boolisManagerX = \fCfalse\fP, boolisManagerZ = \fCfalse\fP, boolisAdministrator = \fCfalse\fP, \fBQObject\fP *parent = \fC0\fP)"

.PP
Définition à la ligne 18 du fichier SH_User\&.cpp\&.
.PP
Références isAdministrator(), isManagerX(), isManagerZ(), isReceptionist(), m_administrator, m_managerX, m_managerZ, m_receptionist, setID(), et setName()\&.
.PP
Référencé par logIn()\&.
.PP
.nf
19     : QObject(parent), m_id(0)
20 {
21     this->setName(name);
22     this->setID(id);
23     this->m_receptionist = isReceptionist;
24     this->m_managerX = isManagerX;
25     this->m_managerZ = isManagerZ;
26     this->m_administrator = isAdministrator;
27 }
.fi
.SH "Documentation des fonctions membres"
.PP 
.SS "SH_User::exists (QVariantlogin)\fC [inline]\fP, \fC [static]\fP, \fC [slot]\fP"

.PP
\fBParamètres:\fP
.RS 4
\fIlogin\fP 
.RE
.PP
\fBRenvoie:\fP
.RS 4
QVariant 
.RE
.PP

.PP
Définition à la ligne 147 du fichier SH_User\&.h\&.
.PP
Références traineeExists(), et userExists()\&.
.PP
Référencé par SH_ApplicationCore::setMode(), et SH_ApplicationCore::userExists()\&.
.PP
.nf
147 {return QVariant(SH_User::userExists(login\&.toString()) || SH_User::traineeExists(login\&.toString()));}
.fi
.SS "int SH_User::id () const\fC [inline]\fP"

.PP
Définition à la ligne 57 du fichier SH_User\&.h\&.
.PP
Références m_id\&.
.PP
Référencé par setID()\&.
.PP
.nf
57 { return this->m_id; }
.fi
.SS "SH_User::isAdministrator () const\fC [inline]\fP"

.PP
\fBRenvoie:\fP
.RS 4
bool 
.RE
.PP

.PP
Définition à la ligne 89 du fichier SH_User\&.h\&.
.PP
Références m_administrator\&.
.PP
Référencé par save(), SH_ApplicationCore::setMode(), et SH_User()\&.
.PP
.nf
89 { return this->m_administrator; }
.fi
.SS "SH_User::isManagerX () const\fC [inline]\fP"

.PP
\fBRenvoie:\fP
.RS 4
bool 
.RE
.PP

.PP
Définition à la ligne 73 du fichier SH_User\&.h\&.
.PP
Références m_managerX\&.
.PP
Référencé par save(), SH_ApplicationCore::setMode(), et SH_User()\&.
.PP
.nf
73 { return this->m_managerX; }
.fi
.SS "SH_User::isManagerZ () const\fC [inline]\fP"

.PP
\fBRenvoie:\fP
.RS 4
bool 
.RE
.PP

.PP
Définition à la ligne 81 du fichier SH_User\&.h\&.
.PP
Références m_managerZ\&.
.PP
Référencé par save(), SH_ApplicationCore::setMode(), et SH_User()\&.
.PP
.nf
81 { return this->m_managerZ; }
.fi
.SS "SH_User::isReceptionist () const"

.PP
\fBRenvoie:\fP
.RS 4
bool 
.RE
.PP

.PP
Définition à la ligne 67 du fichier SH_User\&.cpp\&.
.PP
Références m_receptionist\&.
.PP
Référencé par save(), SH_ApplicationCore::setMode(), et SH_User()\&.
.PP
.nf
68 {
69     return this->m_receptionist;
70 }
.fi
.SS "SH_User::isValid () const"

.PP
\fBRenvoie:\fP
.RS 4
bool 
.RE
.PP

.PP
Définition à la ligne 35 du fichier SH_User\&.cpp\&.
.PP
Références m_id, et m_name\&.
.PP
Référencé par logIn(), SH_ApplicationCore::setUser(), et SH_ApplicationCore::userLogOut()\&.
.PP
.nf
35                             {
36     return ((!this->m_name\&.isEmpty()) && (this->m_id != 0));
37 }
.fi
.SS "SH_User::logIn (QStringlogin, QStringpass)\fC [static]\fP"

.PP
\fBParamètres:\fP
.RS 4
\fIlogin\fP 
.br
\fIpass\fP 
.RE
.PP
\fBRenvoie:\fP
.RS 4
User 
.RE
.PP

.PP
Définition à la ligne 157 du fichier SH_User\&.cpp\&.
.PP
Références SH_DatabaseManager::execSelectQuery(), SH_DatabaseManager::getInstance(), isValid(), SH_User(), traineeExists(), et userExists()\&.
.PP
Référencé par SH_ApplicationCore::setUser()\&.
.PP
.nf
158 {
159     //SH_MessageManager::debugMessage("log in");
160     bool isValid = false;
161     QCryptographicHash encPass(QCryptographicHash::Sha512);
162     encPass\&.addData(pass\&.toUtf8());
163     bool trainee=false;
164     QStringList fields;
165     QString table;
166     if(userExists(login)) {
167         fields << "ID" << "LOGIN" << "ISRECEPTIONIST" << "ISMANAGERX" << "ISMANAGERZ" << "ISADMINISTRATOR";
168         table ="USERS";
169     } else if(traineeExists(login)) {
170         fields << "ID" << "LOGIN";
171         table ="TRAINEES";
172         trainee=true;
173     }
174     QSqlQuery result = SH_DatabaseManager::getInstance()->execSelectQuery(table,fields,"LOGIN='"+login+"' AND ENCRYPTEDPASS='"+QString::fromLatin1(encPass\&.result()\&.toHex())\&.toUpper()+"'");
175     if(result\&.next()) {
176         QSqlRecord rec = result\&.record();
177         if(rec\&.isEmpty() || !result\&.isValid()) {
178             isValid = false;
179         } else {
180             isValid = (rec\&.value(rec\&.indexOf("LOGIN"))\&.toString() == login);
181         }
182 
183         if(isValid) {
184             if(trainee) {
185                 return new SH_Trainee(rec\&.value(rec\&.indexOf("LOGIN"))\&.toString(),rec\&.value(rec\&.indexOf("ID"))\&.toInt());
186             } else {
187                 return new SH_User(rec\&.value(rec\&.indexOf("LOGIN"))\&.toString(),rec\&.value(rec\&.indexOf("ID"))\&.toInt(),(rec\&.value(rec\&.indexOf("ISRECEPTIONIST"))\&.toString()=="1"),(rec\&.value(rec\&.indexOf("ISMANAGERX"))\&.toString()=="1"),(rec\&.value(rec\&.indexOf("ISMANAGERZ"))\&.toString()=="1"),(rec\&.value(rec\&.indexOf("ISADMINISTRATOR"))\&.toString()=="1"));
188             }
189         }
190     }
191     return new SH_User();
192 }
.fi
.SS "QString SH_User::name () const"

.PP
Référencé par SH_Trainee::save(), save(), et setName()\&.
.SS "SH_User::nameChanged ()\fC [signal]\fP"

.SS "int SH_User::roles () const"

.SS "SH_User::rolesChanged ()\fC [signal]\fP"

.SS "bool SH_User::save (QStringpassword)\fC [virtual]\fP"

.PP
Réimplémentée dans \fBSH_Trainee\fP\&.
.PP
Définition à la ligne 130 du fichier SH_User\&.cpp\&.
.PP
Références SH_DatabaseManager::execReplaceQuery(), SH_DatabaseManager::getInstance(), isAdministrator(), isManagerX(), isManagerZ(), isReceptionist(), et name()\&.
.PP
Référencé par SH_ApplicationCore::saveUser(), et setNewPassword()\&.
.PP
.nf
130                                    {
131     QVariantMap map;
132     if(id() > 0) {
133         map\&.insert("ID",QVariant(this->id()));
134     }
135     map\&.insert("LOGIN",QVariant(this->name()));
136     map\&.insert("ISRECEPTIONIST",QVariant(this->isReceptionist()));
137     map\&.insert("ISMANAGERX",QVariant(this->isManagerX()));
138     map\&.insert("ISMANAGERZ",QVariant(this->isManagerZ()));
139     map\&.insert("ISADMINISTRATOR",QVariant(this->isAdministrator()));
140     QCryptographicHash encPass(QCryptographicHash::Sha512);
141     encPass\&.addData(password\&.toUtf8());
142     map\&.insert("ENCRYPTEDPASS",QVariant(QString::fromLatin1(encPass\&.result()\&.toHex())\&.toUpper()));
143 
144     return SH_DatabaseManager::getInstance()->execReplaceQuery("USERS",map);
145 }
.fi
.SS "SH_User::setID (intid)\fC [private]\fP"

.PP
\fBParamètres:\fP
.RS 4
\fIid\fP 
.RE
.PP

.PP
Définition à la ligne 102 du fichier SH_User\&.cpp\&.
.PP
Références id(), et m_id\&.
.PP
Référencé par SH_User()\&.
.PP
.nf
103 {
104     m_id = id;
105 }
.fi
.SS "SH_User::setName (QStringname)\fC [private]\fP"

.PP
\fBParamètres:\fP
.RS 4
\fIname\fP 
.RE
.PP

.PP
Définition à la ligne 45 du fichier SH_User\&.cpp\&.
.PP
Références m_name, et name()\&.
.PP
Référencé par SH_User()\&.
.PP
.nf
46 {
47     m_name = name;
48 }
.fi
.SS "bool SH_User::setNewPassword (QStringnewPass)"

.PP
Définition à la ligne 147 du fichier SH_User\&.cpp\&.
.PP
Références save()\&.
.PP
.nf
147                                             {
148     return this->save(newPass);
149 }
.fi
.SS "SH_User::traineeExists (QStringlogin)\fC [static]\fP"

.PP
\fBParamètres:\fP
.RS 4
\fIlogin\fP 
.RE
.PP
\fBRenvoie:\fP
.RS 4
bool 
.RE
.PP

.PP
Définition à la ligne 124 du fichier SH_User\&.cpp\&.
.PP
Références SH_DatabaseManager::getInstance()\&.
.PP
Référencé par exists(), et logIn()\&.
.PP
.nf
124                                          {
125     //SH_MessageManager::debugMessage("trainee exists");
126     return (SH_DatabaseManager::getInstance()->dataCount("TRAINEES", "LOGIN='"+login+"'") == 1);
127 }
.fi
.SS "SH_User::userExists (QStringlogin)\fC [static]\fP"

.PP
\fBParamètres:\fP
.RS 4
\fIlogin\fP 
.RE
.PP
\fBRenvoie:\fP
.RS 4
bool 
.RE
.PP

.PP
Définition à la ligne 113 du fichier SH_User\&.cpp\&.
.PP
Références SH_DatabaseManager::getInstance()\&.
.PP
Référencé par exists(), et logIn()\&.
.PP
.nf
113                                       {
114     //SH_MessageManager::debugMessage("user exists");
115     return (SH_DatabaseManager::getInstance()->dataCount("USERS", "LOGIN='"+login+"'") == 1);
116 }
.fi
.SS "SH_User::validityChanged ()\fC [signal]\fP"

.SH "Documentation des données membres"
.PP 
.SS "bool SH_User::m_administrator\fC [private]\fP"

.PP
m_administrator 
.PP
Définition à la ligne 213 du fichier SH_User\&.h\&.
.PP
Référencé par isAdministrator(), et SH_User()\&.
.SS "int SH_User::m_id\fC [private]\fP"

.PP
m_id 
.PP
Définition à la ligne 218 du fichier SH_User\&.h\&.
.PP
Référencé par id(), isValid(), et setID()\&.
.SS "bool SH_User::m_managerX\fC [private]\fP"

.PP
m_managerX 
.PP
Définition à la ligne 203 du fichier SH_User\&.h\&.
.PP
Référencé par isManagerX(), et SH_User()\&.
.SS "bool SH_User::m_managerZ\fC [private]\fP"

.PP
m_managerZ 
.PP
Définition à la ligne 208 du fichier SH_User\&.h\&.
.PP
Référencé par isManagerZ(), et SH_User()\&.
.SS "QString SH_User::m_name\fC [private]\fP"

.PP
m_name 
.PP
Définition à la ligne 193 du fichier SH_User\&.h\&.
.PP
Référencé par isValid(), et setName()\&.
.SS "bool SH_User::m_receptionist\fC [private]\fP"

.PP
m_receptionist 
.PP
Définition à la ligne 198 du fichier SH_User\&.h\&.
.PP
Référencé par isReceptionist(), et SH_User()\&.
.SH "Documentation des propriétés"
.PP 
.SS "bool SH_User::administrator\fC [read]\fP"

.PP
Définition à la ligne 23 du fichier SH_User\&.h\&.
.SS "SH_User::id\fC [read]\fP"

.PP
\fBRenvoie:\fP
.RS 4
int 
.RE
.PP

.PP
Définition à la ligne 18 du fichier SH_User\&.h\&.
.PP
Référencé par SH_ApplicationCore::launchServiceCharging()\&.
.SS "bool SH_User::managerX\fC [read]\fP"

.PP
Définition à la ligne 21 du fichier SH_User\&.h\&.
.SS "bool SH_User::managerZ\fC [read]\fP"

.PP
Définition à la ligne 22 du fichier SH_User\&.h\&.
.SS "SH_User::name\fC [read]\fP"

.PP
\fBRenvoie:\fP
.RS 4
QString 
.RE
.PP

.PP
Définition à la ligne 19 du fichier SH_User\&.h\&.
.PP
Référencé par SH_ApplicationCore::setMode(), et SH_ApplicationCore::setUser()\&.
.SS "bool SH_User::receptionist\fC [read]\fP"

.PP
Définition à la ligne 20 du fichier SH_User\&.h\&.
.SS "SH_User::roles\fC [read]\fP"

.PP
\fBRenvoie:\fP
.RS 4
int 
.RE
.PP

.PP
Définition à la ligne 24 du fichier SH_User\&.h\&.
.SS "bool SH_User::valid\fC [read]\fP"

.PP
Définition à la ligne 25 du fichier SH_User\&.h\&.

.SH "Auteur"
.PP 
Généré automatiquement par Doxygen pour PreCheck à partir du code source\&.


src/PreCheck/models/SH_User.cpp
#include <QSqlRecord>
#include <QtCore>
#include <QCryptographicHash>
#include <QDebug>
#include "SH_MessageManager.h"
#include "SH_User.h"
#include "SH_Trainee.h"
#include "SH_DatabaseManager.h"

/*namespace SimplHotel
{*/
/*!
 \details \~french

 \fn SH_User::User

*/
SH_User::SH_User(QString name, int id, bool isReceptionist, bool isManagerX, bool isManagerZ, bool isAdministrator, QObject *parent)
    : QObject(parent), m_id(0), m_receptionist(isReceptionist), m_managerX(isManagerX), m_managerZ(isManagerZ), m_administrator(isAdministrator)
{
    this->setName(name);
    this->setID(id);

    connect(this, &SH_User::receptionistChanged, this, &SH_User::rolesChanged);
    connect(this, &SH_User::managerXChanged, this, &SH_User::rolesChanged);
    connect(this, &SH_User::managerZChanged, this, &SH_User::rolesChanged);
    connect(this, &SH_User::administratorChanged, this, &SH_User::rolesChanged);
}

/*!
 \details \~french

 \fn SH_User::isValid

*/
bool SH_User::isValid() const {
    return ((!this->m_name.isEmpty()) && (this->m_id != 0));
}

/*!
 \details \~french

 \fn SH_User::setName

*/
void SH_User::setName(QString name)
{
    m_name = name;
}

/*!
 \details \~french

 \fn SH_User::roles

*/
int SH_User::roles() const
{
    int nb = 0;
    if(this->isReceptionist()) {
        nb++;
    }
    if(this->isManagerX()) {
        nb++;
    }
    if(this->isManagerZ()) {
        nb++;
    }
    if(this->isAdministrator()) {
        nb++;
    }
    return nb;
}

/*!
 \details \~french

 \fn SH_User::setID

*/
void SH_User::setID(int id)
{
    m_id = id;
}

/*!
 \details \~french

 \fn SH_User::userExists

*/
bool SH_User::userExists(QString login) {
    //SH_MessageManager::debugMessage("user exists");
    return (SH_DatabaseManager::getInstance()->dataCount("USERS", "LOGIN='"+login+"'") == 1);
}

/*!
 \details \~french

 \fn SH_User::traineeExists

*/
bool SH_User::traineeExists(QString login) {
    //SH_MessageManager::debugMessage("trainee exists");
    return (SH_DatabaseManager::getInstance()->dataCount("TRAINEES", "LOGIN='"+login+"'") == 1);
}


bool SH_User::save(QString password) {
    QVariantMap map;
    if(id() > 0) {
        map.insert("ID",QVariant(this->id()));
    }
    map.insert("LOGIN",QVariant(this->name()));
    map.insert("ISRECEPTIONIST",QVariant(this->isReceptionist()));
    map.insert("ISMANAGERX",QVariant(this->isManagerX()));
    map.insert("ISMANAGERZ",QVariant(this->isManagerZ()));
    map.insert("ISADMINISTRATOR",QVariant(this->isAdministrator()));
    QCryptographicHash encPass(QCryptographicHash::Sha512);
    encPass.addData(password.toUtf8());
    map.insert("ENCRYPTEDPASS",QVariant(QString::fromLatin1(encPass.result().toHex()).toUpper()));

    return SH_DatabaseManager::getInstance()->execReplaceQuery("USERS",map);
}

bool SH_User::setNewPassword(QString newPass) {
    return this->save(newPass);
}

/*!
 \details \~french

 \fn SH_User::logIn

*/
SH_User *SH_User::logIn(QString login, QString pass)
{
    //SH_MessageManager::debugMessage("log in");
    bool isValid = false;
    QCryptographicHash encPass(QCryptographicHash::Sha512);
    encPass.addData(pass.toUtf8());
    bool trainee=false;
    QStringList fields;
    QString table;
    if(userExists(login)) {
        fields << "ID" << "LOGIN" << "ISRECEPTIONIST" << "ISMANAGERX" << "ISMANAGERZ" << "ISADMINISTRATOR";
        table ="USERS";
    } else if(traineeExists(login)) {
        fields << "ID" << "LOGIN";
        table ="TRAINEES";
        trainee=true;
    }
    QSqlQuery result = SH_DatabaseManager::getInstance()->execSelectQuery(table,fields,"LOGIN='"+login+"' AND ENCRYPTEDPASS='"+QString::fromLatin1(encPass.result().toHex()).toUpper()+"'");
    if(result.next()) {
        QSqlRecord rec = result.record();
        if(rec.isEmpty() || !result.isValid()) {
            isValid = false;
        } else {
            isValid = (rec.value(rec.indexOf("LOGIN")).toString() == login);
        }

        if(isValid) {
            if(trainee) {
                return new SH_Trainee(rec.value(rec.indexOf("LOGIN")).toString(),rec.value(rec.indexOf("ID")).toInt());
            } else {
                return new SH_User(rec.value(rec.indexOf("LOGIN")).toString(),rec.value(rec.indexOf("ID")).toInt(),(rec.value(rec.indexOf("ISRECEPTIONIST")).toString()=="1"),(rec.value(rec.indexOf("ISMANAGERX")).toString()=="1"),(rec.value(rec.indexOf("ISMANAGERZ")).toString()=="1"),(rec.value(rec.indexOf("ISADMINISTRATOR")).toString()=="1"));
            }
        }
    }
    return new SH_User();
}
/*}*/
